Шаблонизируем проект

MarcoDolya opened the pull request:
Добавлены шаблоны- data, functions, layout, main.
Добавлены функции фильтрации входных данных.

---
:mortar_board: [Шаблонизируем проект](https://up.htmlacademy.ru/php/12/user/1626261/tasks/7)

MarcoDolya committed:
Шаблонизируем проект

Добавлены шаблоны- data, functions, layout, main.
Добавленны функции фильтрации входных данных

oegir commented on templates/main.php:
```diff
@@ -0,0 +1,43 @@
+                </div>
+                <div class="lot__info">
+                    <span class="lot__category"><?= htmlspecialchars($value['categories']); ?></span>
+                    <h3 class="lot__title"><a class="text-link" href="pages/lot.html"><?= $value['name']; ?></a></h3>
```
> Считайте, что содержимое главной страницы (список категорий и объявлений) получено от пользователя, поэтому его нужно соответствующим образом фильтровать для защиты от XSS.

По-видимому под списком объявлений подразумеваются наименования лотов, их нужно обернуть в `htmlspecialchars()`:
```
<?= htmlspecialchars($value['name']); ?>
```

oegir commented on templates/main.php:
```diff
@@ -0,0 +1,43 @@
+        <?php foreach ($lots as $key => $value): ?>
+            <li class="lots__item lot">
+                <div class="lot__image">
+                    <img src="<?= $value['image']; ?>" width="350" height="260" alt="<?= $value['name']; ?>">
```
Тег `img` можно "сломать" если в начало строки с именем вставить символы `">`. Это будет воспринято браузером как закрытие тега `img` и злоумышленник сможет далее добавить на страницу свой скрипт. Поэтому здесь подстановку имени также стоило обернуть в `htmlspecialchars()`
```
<?= htmlspecialchars($value['name']); ?>
```

oegir replied:
Нашел любопытный [пример](https://habr.com/ru/sandbox/90823/) внедрение XSS-кода в имени картинки, так что и имя файла тоже можно на всякий случай обернуть в `htmlspecialchars()`
```
<img src="<?= htmlspecialchars($value['image']); ?>"
```

MarcoDolya committed:
фильтрация

Работала бы такая конструкция? -
formatSum(htmlspecialchairs($value['price']))?

oegir committed:
Merge branch 'master' into module3-task1
